src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,651 @@
-
-            inner.source_upper.advance(pid, diff);
-        }
+        // Ensure we only add consolidated batches to our trace
+        consolidation::consolidate_updates(&mut batch.updates);
+        inner.remap_trace.extend(batch.updates.iter().cloned());
+        inner.source_upper.update_iter(
+            batch
+                .updates
+                .into_iter()
+                .map(|(src_ts, _ts, diff)| (src_ts, diff)),
+        );
         inner.upper = batch.upper;
+        if !inner.initialized && PartialOrder::less_equal(&inner.as_of, &inner.upper) {
+            inner.initialized = true;
+            let as_of = inner.as_of.clone();
+            drop(inner);
+            self.compact(as_of);
+        }
     }
 
-    /// Reclocks a batch of messages timestamped with `SourceTime` and returns an iterator of
-    /// messages timestamped with `DestTime`.
-    ///
-    /// The returned iterator will drain the provided batch as it being consumed. It is not
-    /// guaranteed that the provided batch will be cleared if the iterator is dropped in a
-    /// partially consumed state.
+    /// Reclocks a batch of messages timestamped with `FromTime` and returns an iterator of
+    /// messages timestamped with `IntoTime`.
     ///
-    /// The method returns an error if any of the messages is timestamped at a `SourceTime` that is
-    /// not beyond the since frontier. The error will contain the offending `SourceTime`.
+    /// Each item of the resulting iterator will be associated with either the time it should be
+    /// reclocked to or an error indicating that a reclocking decision could not be taken with the
+    /// data that we have at hand.
     ///
-    /// This method returns `None` if we don't yet have enough bindings to cover
-    /// this batch, that is if the frontier in `SourceTime` is not yet advanced
-    /// far enough.
-    pub fn reclock<'a, M>(
+    /// This method is most efficient when the to be reclocked iterator presents data in contiguous
+    /// runs with the same `FromTime`.
+    // This is dead code until we get rid of the compatibility layer
+    #[allow(dead_code)]
+    pub fn reclock<'a, M: 'a>(
         &'a self,
-        batch: &'a mut HashMap<PartitionId, Vec<(M, MzOffset)>>,
-    ) -> Result<Option<ReclockIter<'a, M>>, (PartitionId, MzOffset)> {
-        let inner = RefCell::borrow(&self.inner);
-
-        let mut batch_upper = HashMap::with_capacity(batch.len());
-        for (pid, messages) in batch.iter_mut() {
-            messages.sort_unstable_by(|a, b| a.1.cmp(&b.1));
-            if let Some((_msg, offset)) = messages.first() {
-                let part_since = inner.partition_since(pid);
-                if !(part_since <= *offset) {
-                    return Err((pid.clone(), *offset));
+        batch: impl IntoIterator<Item = (M, FromTime)> + 'a,
+    ) -> impl Iterator<Item = (M, Result<IntoTime, ReclockError<FromTime>>)> + 'a
+    where
+        IntoTime: TotalOrder,
```

Comment:
can we just put this bound on the entire impl block?

Reply:
No because `reclock_time` doesn't require `IntoTime`